File: go/pkg/jd/chunkcoords/chunkcoords.go
// Package chunkcoords — Coordenadas de chunk (floor division, keys, radio).
package chunkcoords

import (
	"fmt"
	"math"
)

// CoordFromCell returns chunk index for a cell coordinate (floor division, negative-safe).
func CoordFromCell(cell, chunkSize int) int {
	if cell >= 0 {
		return cell / chunkSize
	}
	return (cell - chunkSize + 1) / chunkSize
}

// Key returns stable "cx,cy" string.
func Key(cx, cy int) string {
	return fmt.Sprintf("%d,%d", cx, cy)
}

// KeyFromCell returns chunk key containing cell (x, y).
func KeyFromCell(x, y, chunkSize int) string {
	return Key(CoordFromCell(x, chunkSize), CoordFromCell(y, chunkSize))
}

// CellBounds returns inclusive (xMin, xMax, yMin, yMax) for chunk tile.
func CellBounds(cx, cy, chunkSize int) (xMin, xMax, yMin, yMax int) {
	xMin = cx * chunkSize
	yMin = cy * chunkSize
	return xMin, xMin + chunkSize - 1, yMin, yMin + chunkSize - 1
}

// KeysInRadius returns chunk keys intersecting a disc (conservative bbox).
func KeysInRadius(centerX, centerY float64, radiusCells, chunkSize int) map[string]struct{} {
	cx := int(math.Floor(centerX))
	cy := int(math.Floor(centerY))
	minX, maxX := cx-radiusCells, cx+radiusCells
	minY, maxY := cy-radiusCells, cy+radiusCells
	cx0 := CoordFromCell(minX, chunkSize)
	cx1 := CoordFromCell(maxX, chunkSize)
	cy0 := CoordFromCell(minY, chunkSize)
	cy1 := CoordFromCell(maxY, chunkSize)
	out := make(map[string]struct{})
	for i := cx0; i <= cx1; i++ {
		for j := cy0; j <= cy1; j++ {
			out[Key(i, j)] = struct{}{}
		}
	}
	return out
}

// UnionForPlayers merges chunk windows for all player positions.
func UnionForPlayers(positions [][2]float64, radiusCells, chunkSize int) map[string]struct{} {
	out := make(map[string]struct{})
	for _, p := range positions {
		for k := range KeysInRadius(p[0], p[1], radiusCells, chunkSize) {
			out[k] = struct{}{}
		}
	}
	return out
}

File: rust/terrain-service/src/chunkcoords.rs
//! Chunk coordinates — mirror go/pkg/jd/chunkcoords and Python chunk_coords.py.

pub fn coord_from_cell(cell: i32, chunk_size: i32) -> i32 {
    if cell >= 0 {
        cell / chunk_size
    } else {
        (cell - chunk_size + 1) / chunk_size
    }
}

/// `key` — operación pública del módulo.
pub fn key(cx: i32, cy: i32) -> String {
    format!("{cx},{cy}")
}

/// `key_from_cell` — operación pública del módulo.
pub fn key_from_cell(x: i32, y: i32, chunk_size: i32) -> String {
    key(
        coord_from_cell(x, chunk_size),
        coord_from_cell(y, chunk_size),
    )
}

/// `cell_bounds` — operación pública del módulo.
pub fn cell_bounds(cx: i32, cy: i32, chunk_size: i32) -> (i32, i32, i32, i32) {
    let x_min = cx * chunk_size;
    let y_min = cy * chunk_size;
    (
        x_min,
        x_min + chunk_size - 1,
        y_min,
        y_min + chunk_size - 1,
    )
}
